### DISCUSSION TITLE
Mini-comment commentstring autocmd

### DISCUSSION THREAD

**Discussion by krishna-bala** [LazyVim/LazyVim#680] [Recipes]
Thought I would share this as there weren't many discussions here/elsewhere. The default `commentstring` for C/C++ files is `/* %s */`. `commentstring` is what `mini.comment` uses when adding/removing comments. In order to change the `commentstring` for C/C++ files, I added the following to `lua/config/autocmds.lua`:

```
-- lua/config/autocmds.lua
-- Autocmds are automatically loaded on the VeryLazy event
-- Default autocmds that are always set: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/autocmds.lua
-- Add any additional autocmds here

local function augroup(name)
  return vim.api.nvim_create_augroup("lazyvim_config_" .. name, { clear = true })
end

vim.api.nvim_create_autocmd("FileType", {
  group = augroup("commentstring"),
  pattern = { "*.cc", "*.c", "*.h" },
  callback = function()
    vim.opt.commentstring = "// %s"
  end,
  desc = "Change commentstring for c/c++ files",
})
```

**Comment by folke** [comment]
You shouldn't need to set this manually, since we use https://github.com/JoosepAlviste/nvim-ts-context-commentstring for that. Were you experiencing any issues with commentstring?

**Reply by krishna-bala** [reply]
No issues with the default. I was trying to change the commentstring for C/C++ files from `/* %s */` to `// %s` as that is the style my company uses. While I achieved the desired results using the method above, your comment prompted me to look at [:h ts-context-commentstring](https://github.com/JoosepAlviste/nvim-ts-context-commentstring/blob/main/doc/nvim-ts-context-commentstring.txt) to see if I can configure this behavior without the `autocmd` hack above. That doc has some instructions on how to change the `commentstring` for a given language using `nvim-treesitter`. Here's the example snippet the help doc provided:
```
Support for more languages can be added quite easily by passing a `config` table
when configuring the plugin:
>lua
  require('nvim-treesitter.configs').setup {
    context_commentstring = {
      enable = true,
      config = {
        css = '// %s',
      },
    },
  }
<
``` 
However, I'm having some trouble using this. It looks like a modification of the `nvim-treesitter` opts, so I added the following to my `lua/plugins/nvim-treesitter.lua`:

```lua
-- lua/plugins/nvim-treesitter.lua
return {
  "nvim-treesitter/nvim-treesitter",
  dependencies = {
    "JoosepAlviste/nvim-ts-context-commentstring",
  },
  opts = function(_, opts)
    opts.context_commentstring = vim.tbl_extend("force", opts.context_commentstring, {
      enable = true,
      config = {
        cpp = "// %s",
      },
    })
  end,
}
```

But it doesn't seem to update the commentstring as I would expect (commentstring is still `/* %s */`)

I also tried the following with no change in behavior:
```lua
-- lua/plugins/nvim-treesitter.lua
return {
  "nvim-treesitter/nvim-treesitter",
  dependencies = {
    "JoosepAlviste/nvim-ts-context-commentstring",
    config = function()
      require("nvim-treesitter.configs").setup({
        context_commentstring = {
          enable = true,
          config = {
            c = "// %s",
            cpp = "// %s",
          },
        },
      })
    end,
  },
}
```

Any insight appreciated, but since I have a working solution (`autocmd`) this isn't really an issue for me.

**Reply by Jomik** [reply]
try something like
```lua
-- lua/plugins/nvim-treesitter.lua
return {
  "nvim-treesitter/nvim-treesitter",
  dependencies = {
    { "JoosepAlviste/nvim-ts-context-commentstring", lazy = true },
  },
  opts = {
    context_commentstring = {
      enable = true,
      config = {
        c = "// %s",
        cpp = "// %s",
      },
    },
  },
}
```
Unsure if there is a better way of adding to the commentstring options than this, as it is initialised through treesitter.

**Reply by krishna-bala** [reply]
Unfortunately that didn't work for me either. Not sure if you were able to test this and got different results.

**Reply by dpetka2001** [reply]
For some reason the `context_commentstring` module seems to not be enabled. After I did `TSBufEnable context_commentstring` (or `TSEnable` depending on if you wanted it enabled only for the buffer or globally) the configuration seemed to work. After that I changed `lazy = true` to `lazy = false` and it seems to be working correctly without needing the above commands. It must have to do something with not correctly updating `commentstring` for `mini.comment`? Maybe if @folke added it as a dependency to `mini.comment` the problem could be avoided? Or like my solution it should not be `lazy = true`.

Edit: I used @Jomik 's configuration as reference and it worked after i changed `lazy` to `false`.

**Reply by krishna-bala** [reply]
Hmm this didn't seem to work for me despite trying your suggestions (and other combinations of `TSBufEnable`, `TSEnable`, and `lazy = true/false`).

Not sure if this helps, but if I remove everything from my `lua/plugins/nvim-treesitter.lua` like so
```lua
-- lua/plugins/nvim-treesitter.lua VERSION A
if true then
  return {}
end
```

and start typing `:TSEnable context_commentstring`, it doesn't have a completion (as expected). If I submit that command then call `:TSModuleInfo`, I don't see `context_commentstring` in the displayed output (again, expected). 

When I change the file to 
```lua
-- lua/plugins/nvim-treesitter.lua VERSION B
return {
  "nvim-treesitter/nvim-treesitter",
  dependencies = {
    { "JoosepAlviste/nvim-ts-context-commentstring", lazy = false },
  },
  opts = {
    context_commentstring = {
      enable = true,
      config = {
        c = "// %s",
        cpp = "// %s",
      },
    },
  },
}
```
and start typing `:TSEnable context_commentstring`, the completion shows up. If I submit that command and then call `:TSModuleInfo`, I am able to see `context_commentstring` in the output:
```
VERSION B :TSModuleInfo OUTPUT

>> default         context_commentstring  highlight  illuminate  incremental_selection  indent  
ada                ✗                      ✗          ✗           ✗                      ✗       
agda               ✗                      ✗          ✗           ✗                      ✗       
arduino            ✗                      ✗          ✗           ✗                      ✗       
astro              ✗                      ✗          ✗           ✗                      ✗       
awk                ✗                      ✗          ✗           ✗                      ✗       
.
.
.
c                  ✓                      ✓          ✓           ✓                      ✓       
.
.
.
cpp                ✗                      ✗          ✗           ✗                      ✗       
.
.
.
```

So, it seems like it is able to detect that I'm loading `nvim-ts-context-commentstring` but it's not receiving the options that I set (cpp does not have a check mark, and `:set cms?` returns `/* %s */`). 

I can confirm that this is the same behavior as using the default `nvim-ts-context-commenstring` initialization. If I change the `lua/plugins/nvim-treesitter.lua` again, this time modifying VERSION B and removing the `opts` table like so:
```lua
-- lua/plugins/nvim-treesitter.lua VERSION C
return {
  "nvim-treesitter/nvim-treesitter",
  dependencies = {
    { "JoosepAlviste/nvim-ts-context-commentstring", lazy = false },
  },
}
```
and call `TSModuleInfo`, I get the exact same output as above (`VERSION B :TSModuleInfo OUTPUT`).

Hope this makes sense (and if it doesn't feel free to ignore!)

**Reply by dpetka2001** [reply]
Strange, I can't seem to be able to replicate this. I used the VERSION B config and there's no need for extra commands. `TSModuleInfo` shows `context_commentstring` enabled for both c/cpp and `set cms?` shows `commentsting=// %s` inside a cpp file. Did you restart your Neovim after the changes in your config?

**Reply by folke** [reply]
Did you install the cpp TS parser?

**Reply by folke** [reply]
ts-context-commentstring needs the correct ts parser to work. The output from mod info clearly shows you dont have it installed.

**Reply by dpetka2001** [reply]
From your output, maybe I'm wrong perhaps, but are you sure you have `cpp` parser installed in treesitter?

**Reply by folke** [reply]
exactly

**Reply by dpetka2001** [reply]
This still doesn't solve the problem of correctly updating the `commentstring` @folke. Without `lazy=false` it doesn't seem to get updated from the treesitter config. Maybe I'm wrong, but could you perhaps look into that?

**Reply by krishna-bala** [reply]
Wow, I thought that was one of the first things I did... not sure how I didn't catch that. Sorry for the trouble. After installing, the implementation suggested by @Jomik works for me.

Thanks for the feedback.

**Reply by folke** [reply]
This works for me:

```lua
  {
    "nvim-treesitter/nvim-treesitter",
    dependencies = { "JoosepAlviste/nvim-ts-context-commentstring" },
    opts = {
      context_commentstring = {
        config = {
          lua = "|| %s",
        },
      },
    },
  },
```

The reason I didn't add it as a dep for treesitter is because I only want to have it loaded when doing commenting.

This means that commenstring loads after treesitter, and looking at its code, it basically overwrites the user settings with the defaults...

**Reply by dpetka2001** [reply]
Yes, right now that I tested again without `lazy = false`, it seems to work just fine. Don't know what went wrong in my first test and was under the impression that `lazy` was the culprit. Sorry for that.

**Comment by bonjormond** [comment]
For anyone who's tried OP's solution and  had it not detect filetypes properly, try changing

`  pattern = { "*.cpp", "*.cc", "*.c", "*.h" },`
to this:
`  pattern = { "cpp", "cc", "c", "h" },`
which worked for me (even though :h api.txt shows OP's to be correct.)
Super weird, hopefully that helps someone else though.
